<p style="text-align:center">
    <a href="https://skills.network" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>


# **Launch Sites Locations Analysis with Folium**


Estimated time needed: **40** minutes


The launch success rate may depend on many factors such as payload mass, orbit type, and so on. It may also depend on the location and proximities of a launch site, i.e., the initial position of rocket trajectories. Finding an optimal location for building a launch site certainly involves many factors and hopefully we could discover some of the factors by analyzing the existing launch site locations.


In the previous exploratory data analysis labs, you have visualized the SpaceX launch dataset using `matplotlib` and `seaborn` and discovered some preliminary correlations between the launch site and success rates. In this lab, you will be performing more interactive visual analytics using `Folium`.


## Objectives


This lab contains the following tasks:
- **TASK 1:** Mark all launch sites on a map
- **TASK 2:** Mark the success/failed launches for each site on the map
- **TASK 3:** Calculate the distances between a launch site to its proximities

After completed the above tasks, you should be able to find some geographical patterns about launch sites.


Let's first import required Python packages for this lab:


In [1]:
!pip3 install folium
!pip3 install wget
!pip3 install pandas

In [2]:
import folium
import wget
import pandas as pd

In [3]:
# Import folium MarkerCluster plugin
from folium.plugins import MarkerCluster
# Import folium MousePosition plugin
from folium.plugins import MousePosition
# Import folium DivIcon plugin
from folium.features import DivIcon

If you need to refresh your memory about folium, you may download and refer to this previous folium lab:


[Generating Maps with Python](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/DV0101EN-3-5-1-Generating-Maps-in-Python-py-v2.0.ipynb)


## Task 1: Mark all launch sites on a map


First, let's try to add each site's location on a map using site's latitude and longitude coordinates


The following dataset with the name `spacex_launch_geo.csv` is an augmented dataset with latitude and longitude added for each site.


In [4]:
# Download and read the `spacex_launch_geo.csv`
spacex_csv_file = wget.download('https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/datasets/spacex_launch_geo.csv')
spacex_df=pd.read_csv(spacex_csv_file)

Now, you can take a look at what are the coordinates for each site.


In [5]:
# Select relevant sub-columns: `Launch Site`, `Lat(Latitude)`, `Long(Longitude)`, `class`
spacex_df = spacex_df[['Launch Site', 'Lat', 'Long', 'class']]
launch_sites_df = spacex_df.groupby(['Launch Site'], as_index=False).first()
launch_sites_df = launch_sites_df[['Launch Site', 'Lat', 'Long']]
launch_sites_df

Launch Site        Lat        Long
0   CCAFS LC-40  28.562302  -80.577356
1  CCAFS SLC-40  28.563197  -80.576820
2    KSC LC-39A  28.573255  -80.646895
3   VAFB SLC-4E  34.632834 -120.610745

Above coordinates are just plain numbers that can not give you any intuitive insights about where are those launch sites. If you are very good at geography, you can interpret those numbers directly in your mind. If not, that's fine too. Let's visualize those locations by pinning them on a map.


We first need to create a folium `Map` object, with an initial center location to be NASA Johnson Space Center at Houston, Texas.


In [6]:
# Start location is NASA Johnson Space Center
nasa_coordinate = [29.559684888503615, -95.0830971930759]
site_map = folium.Map(location=nasa_coordinate, zoom_start=10)

We could use `folium.Circle` to add a highlighted circle area with a text label on a specific coordinate. For example,


In [7]:
# Create a blue circle at NASA Johnson Space Center's coordinate with a popup label showing its name
circle = folium.Circle(nasa_coordinate, radius=1000, color='#d35400', fill=True).add_child(folium.Popup('NASA Johnson Space Center'))
# Create a blue circle at NASA Johnson Space Center's coordinate with a icon showing its name
marker = folium.map.Marker(
    nasa_coordinate,
    # Create an icon as a text label
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'NASA JSC',
        )
    )
site_map.add_child(circle)
site_map.add_child(marker)

and you should find a small yellow circle near the city of Houston and you can zoom-in to see a larger circle.


Now, let's add a circle for each launch site in data frame `launch_sites`


_TODO:_  Create and add `folium.Circle` and `folium.Marker` for each launch site on the site map


An example of folium.Circle:


`folium.Circle(coordinate, radius=1000, color='#000000', fill=True).add_child(folium.Popup(...))`


An example of folium.Marker:


`folium.map.Marker(coordinate, icon=DivIcon(icon_size=(20,20),icon_anchor=(0,0), html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'label', ))`


In [8]:
# Initialize the map
site_map = folium.Map(location=nasa_coordinate, zoom_start=5)

# For each launch site, add a Circle and a Marker with a label
for index, row in launch_sites_df.iterrows():
    # Add a circle
    folium.Circle(
        location=[row['Lat'], row['Long']],
        radius=1000,
        color='#d35400',  # brighter orange for visibility
        fill=True
    ).add_child(folium.Popup(row['Launch Site'])).add_to(site_map)

    # Add a marker with label
    folium.map.Marker(
        [row['Lat'], row['Long']],
        icon=DivIcon(
            icon_size=(20,20),
            icon_anchor=(0,0),
            html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % row['Launch Site'],
        )
    ).add_to(site_map)

site_map


The generated map with marked launch sites should look similar to the following:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_markers.png">
</center>


Now, you can explore the map by zoom-in/out the marked areas
, and try to answer the following questions:
- Are all launch sites in proximity to the Equator line?
- Are all launch sites in very close proximity to the coast?

Also please try to explain your findings.


# Explanation of Findings
Alright! Let’s analyze the map I just created with the launch sites.

**1. Proximity to the Equator:**

* The Equator is at latitude 0°.
* Looking at my launch sites:

  * **CCAFS LC-40:** 28.5623° N
  * **CCAFS SLC-40:** 28.5632° N
  * **KSC LC-39A:** 28.5733° N
  * **VAFB SLC-4E:** 34.6328° N
* **Observation:** All launch sites are north of the Equator, roughly between 28°N and 34°N.
* **Conclusion:** They are not exactly on the Equator but are relatively close in terms of latitude (tropical/subtropical zone), which is beneficial for rocket launches because being closer to the Equator gives a small speed boost from Earth's rotation.

**2. Proximity to the coast:**

* All Florida sites (**CCAFS LC-40, CCAFS SLC-40, KSC LC-39A**) are extremely close to the Atlantic Ocean coast. You can literally see the coastline from  zoomed-in map.
* **VAFB SLC-4E** in California is close to the Pacific coast.
* **Conclusion:** All launch sites are very close to the coast. This is intentional—launching over water reduces risk to populated areas in case of a failure.

**Summary of Findings:**

* **Equator proximity:** Not exactly on the Equator but relatively near, which is ideal for launches.
* **Coast proximity:** All sites are near the coast, maximizing safety and logistics efficiency.




# Task 2: Mark the success/failed launches for each site on the map


Next, let's try to enhance the map by adding the launch outcomes for each site, and see which sites have high success rates.
Recall that data frame spacex_df has detailed launch records, and the `class` column indicates if this launch was successful or not


In [9]:
spacex_df.tail(10)

Launch Site        Lat       Long  class
46    KSC LC-39A  28.573255 -80.646895      1
47    KSC LC-39A  28.573255 -80.646895      1
48    KSC LC-39A  28.573255 -80.646895      1
49  CCAFS SLC-40  28.563197 -80.576820      1
50  CCAFS SLC-40  28.563197 -80.576820      1
51  CCAFS SLC-40  28.563197 -80.576820      0
52  CCAFS SLC-40  28.563197 -80.576820      0
53  CCAFS SLC-40  28.563197 -80.576820      0
54  CCAFS SLC-40  28.563197 -80.576820      1
55  CCAFS SLC-40  28.563197 -80.576820      0

Next, let's create markers for all launch records.
If a launch was successful `(class=1)`, then we use a green marker and if a launch was failed, we use a red marker `(class=0)`


Note that a launch only happens in one of the four launch sites, which means many launch records will have the exact same coordinate. Marker clusters can be a good way to simplify a map containing many markers having the same coordinate.


Let's first create a `MarkerCluster` object


In [10]:
marker_cluster = MarkerCluster()


_TODO:_ Create a new column in `launch_sites` dataframe called `marker_color` to store the marker colors based on the `class` value


In [ ]:

# Apply a function to check the value of `class` column
# If class=1, marker_color value will be green
# If class=0, marker_color value will be red


In [11]:
# Function to assign color to launch outcome
def assign_marker_color(launch_outcome):
    if launch_outcome == 1:
        return 'green'
    else:
        return 'red'

spacex_df['marker_color'] = spacex_df['class'].apply(assign_marker_color)
spacex_df.tail(10)

Launch Site        Lat       Long  class marker_color
46    KSC LC-39A  28.573255 -80.646895      1        green
47    KSC LC-39A  28.573255 -80.646895      1        green
48    KSC LC-39A  28.573255 -80.646895      1        green
49  CCAFS SLC-40  28.563197 -80.576820      1        green
50  CCAFS SLC-40  28.563197 -80.576820      1        green
51  CCAFS SLC-40  28.563197 -80.576820      0          red
52  CCAFS SLC-40  28.563197 -80.576820      0          red
53  CCAFS SLC-40  28.563197 -80.576820      0          red
54  CCAFS SLC-40  28.563197 -80.576820      1        green
55  CCAFS SLC-40  28.563197 -80.576820      0          red

_TODO:_ For each launch result in `spacex_df` data frame, add a `folium.Marker` to `marker_cluster`


In [12]:
from folium.features import DivIcon

# Initial the map
site_map = folium.Map(location=nasa_coordinate, zoom_start=5)

# Add launch site circles and markers
for index, row in launch_sites_df.iterrows():
    coordinate = [row['Lat'], row['Long']]

    # Circle (radius in meters, scales with zoom)
    circle = folium.Circle(
        location=coordinate,
        radius=50,  # 1 km
        color='yellow',
        fill=True,
        fill_color='red',
        fill_opacity=0.5,
        popup=folium.Popup(row['Launch Site'])
    )

    # Label marker
    marker = folium.map.Marker(
        coordinate,
        icon=DivIcon(
            icon_size=(20,20),
            icon_anchor=(0,0),
            html=f'<div style="font-size: 12px; color:blue;"><b>{row["Launch Site"]}</b></div>'
        )
    )

    site_map.add_child(circle)
    site_map.add_child(marker)


# Add launch outcome markers using MarkerCluster
marker_cluster = MarkerCluster()
for index, record in spacex_df.iterrows():
    marker = folium.Marker(
        location=[record['Lat'], record['Long']],
        icon=folium.Icon(color='white', icon_color=record['marker_color']),
        popup=(
            f"Launch Site: {record['Launch Site']}<br>"
            f"Outcome: {'Success' if record['class']==1 else 'Failure'}"
        )
    )
    marker_cluster.add_child(marker)

site_map.add_child(marker_cluster)

# Display the map
site_map


Your updated map may look like the following screenshots:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_cluster.png">
</center>


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_cluster_zoomed.png">
</center>


From the color-labeled markers in marker clusters, you should be able to easily identify which launch sites have relatively high success rates.


# TASK 3: Calculate the distances between a launch site to its proximities


Next, we need to explore and analyze the proximities of launch sites.


Let's first add a `MousePosition` on the map to get coordinate for a mouse over a point on the map. As such, while you are exploring the map, you can easily find the coordinates of any points of interests (such as railway)


In [18]:
# Add Mouse Position to get the coordinate (Lat, Long) for a mouse over on the map
formatter = "function(num) {return L.Util.formatNum(num, 5);};"
mouse_position = MousePosition(
    position='topright',
    separator=' Long: ',
    empty_string='NaN',
    lng_first=False,
    num_digits=20,
    prefix='Lat:',
    lat_formatter=formatter,
    lng_formatter=formatter,
)

site_map.add_child(mouse_position)
site_map

Now zoom in to a launch site and explore its proximity to see if you can easily find any railway, highway, coastline, etc. Move your mouse to these points and mark down their coordinates (shown on the top-left) in order to the distance to the launch site.


You can calculate the distance between two points on the map based on their `Lat` and `Long` values using the following method:


In [14]:
from math import sin, cos, sqrt, atan2, radians

def calculate_distance(lat1, lon1, lat2, lon2):
    # approximate radius of earth in km
    R = 6373.0

    lat1 = radians(lat1)
    lon1 = radians(lon1)
    lat2 = radians(lat2)
    lon2 = radians(lon2)

    dlon = lon2 - lon1
    dlat = lat2 - lat1

    a = sin(dlat / 2)**2 + cos(lat1) * cos(lat2) * sin(dlon / 2)**2
    c = 2 * atan2(sqrt(a), sqrt(1 - a))

    distance = R * c
    return distance

_TODO:_ Mark down a point on the closest coastline using MousePosition and calculate the distance between the coastline point and the launch site.


In [15]:
coastline_lat = 28.56367
coastline_lon = -80.56822

# Launch site: CCAFS SLC-40
launch_site_lat = 28.563197
launch_site_lon = -80.576820

# Calculate distance
distance_coastline = calculate_distance(
    launch_site_lat, launch_site_lon,
    coastline_lat, coastline_lon
)

print(f"Distance from CCAFS SLC-40 to closest coastline: {distance_coastline:.2f} km\n\n")


Distance from CCAFS SLC-40 to closest coastline: 0.84 km




_TODO:_ After obtained its coordinate, create a `folium.Marker` to show the distance


In [16]:
distance_marker = folium.Marker(
    location=[coastline_lat, coastline_lon],
    icon=DivIcon(
        icon_size=(80,20),
        icon_anchor=(20, -10),  # <-- moves text so it does NOT overlap
        html='''
        <div style="
            background:white;
            padding:2px 4px;
            border-radius:4px;
            font-size:12px;
            color:#d35400;
            font-weight:bold;
        ">
            {:.2f} KM
        </div>
        '''.format(distance_coastline)
    )
)

site_map.add_child(distance_marker)
site_map


_TODO:_ Draw a `PolyLine` between a launch site to the selected coastline point


In [17]:
# Create a list of coordinates for the line
coordinates = [
    [launch_site_lat, launch_site_lon],
    [coastline_lat, coastline_lon]
]

# Create the PolyLine (weight=1 exactly as instructed)
lines = folium.PolyLine(
    locations=coordinates,
    weight=1,
    color='blue'   # optional, Folium defaults to blue anyway
)

# Add to map
site_map.add_child(lines)

# Display map
site_map


Your updated map with distance line should look like the following screenshot:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_distance.png">
</center>


_TODO:_ Similarly, you can draw a line betwee a launch site to its closest city, railway, highway, etc. You need to use `MousePosition` to find the their coordinates on the map first


A railway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/railway.png">
</center>


A highway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/highway.png">
</center>


A city map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/city.png">
</center>


In [21]:
# Create a marker with distance to a closest city, railway, highway, etc.
# Draw a line between the marker to the launch site
# Launch site (CCAFS SLC-40)
launch_site_lat = 28.563197
launch_site_lon = -80.576820

# Train coordinates
train_lat = 28.47558
train_lon = -80.77793

# Highway coordinates
highway_lat = 28.56284
highway_lon = -80.57071

# City (Melbourne)
city_lat = 28.07592
city_lon = -80.60823


In [22]:
# TRAIN PolyLine
train_line = folium.PolyLine(
    locations=[
        [launch_site_lat, launch_site_lon],
        [train_lat, train_lon]
    ],
    weight=1
)
site_map.add_child(train_line)


In [23]:
# HIGHWAY PolyLine
highway_line = folium.PolyLine(
    locations=[
        [launch_site_lat, launch_site_lon],
        [highway_lat, highway_lon]
    ],
    weight=1
)
site_map.add_child(highway_line)


In [24]:
# CITY PolyLine
city_line = folium.PolyLine(
    locations=[
        [launch_site_lat, launch_site_lon],
        [city_lat, city_lon]
    ],
    weight=1
)
site_map.add_child(city_line)


In [25]:
site_map


After you plot distance lines to the proximities, you can answer the following questions easily:
- Are launch sites in close proximity to railways?
- Are launch sites in close proximity to highways?
- Are launch sites in close proximity to coastline?
- Do launch sites keep certain distance away from cities?

Also please try to explain your findings.


In [30]:
# --- Distance Calculations ---
dist_coast = calculate_distance(launch_site_lat, launch_site_lon, coastline_lat, coastline_lon)
dist_train  = calculate_distance(launch_site_lat, launch_site_lon, train_lat, train_lon)
dist_high  = calculate_distance(launch_site_lat, launch_site_lon, highway_lat, highway_lon)
dist_city  = calculate_distance(launch_site_lat, launch_site_lon, city_lat, city_lon)

# --- Print All Distances ---
print("Distance to coastline: {:.2f} km".format(dist_coast))
print("Distance to railway:   {:.2f} km".format(dist_train))
print("Distance to highway:   {:.2f} km".format(dist_high))
print("Distance to city:      {:.2f} km".format(dist_city))



Distance to coastline: 0.84 km
Distance to railway:   21.94 km
Distance to highway:   0.60 km
Distance to city:      54.29 km


# Explanation of Findings
**Are launch sites in close proximity to railways?**

Yes. Railway point was:

```
Train: lat 28.47558, lon -80.77793
```

The distance is ≈ 21.94 km, which is reasonably close for logistics.

**✔ Explanation**

Railways are useful for transporting large rocket components, but they do not need to be extremely close because heavy parts are usually shipped by road or barge.

****************************
**Are launch sites in close proximity to highways?**

Yes — very close.

Highway point:
```
Highway: lat 28.56284, lon -80.57071
```
Distance from launch site ≈ 0.6 km.

**✔ Explanation**

Launch operations require:

*   frequent movement of equipment
*   emergency access
*   safe evacuation routes

So launch sites are intentionally built near accessible major roads.
****************************
**Are launch sites in close proximity to coastline?**

Yes — very close (usually < 1 km).

Measured coast distance ≈ 0.84 km.

**✔ Explanation**

Launch sites are placed near coastlines because:

*   rockets must launch over the ocean to avoid falling debris hitting populated areas
*   large open water areas provide safe downrange zones coastal areas reduce air traffic and population risk

This is the strongest and most intentional pattern.

*****************************
**Do launch sites keep a certain distance away from cities?**

Yes — they maintain large separation from cities.

The city (Melbourne, FL) coordinate gave a distance of ≈ 54.29 km.

**✔ Explanation**

Launch sites must be isolated from populated areas because:

*   launches can fail

*   explosions, noise, and debris create risk

*   safety regulations require buffer zones of tens of kilometers

*   cities must be far enough to avoid sonic boom and shockwave damage

So being far from cities is planned and required.

# Next Steps:

Now you have discovered many interesting insights related to the launch sites' location using folium, in a very interactive way. Next, you will need to build a dashboard using Ploty Dash on detailed launch records.


## Authors


[Yan Luo](https://www.linkedin.com/in/yan-luo-96288783/)


### Other Contributors


Joseph Santarcangelo


## Change Log


|Date (YYYY-MM-DD)|Version|Changed By|Change Description|
|-|-|-|-|
|2021-05-26|1.0|Yan|Created the initial version|


Copyright © 2021 IBM Corporation. All rights reserved.
